File: r988/D.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <tuple>

using namespace std;

int main() {
    ios::sync_with_stdio(false); // Disable synchronization with C-style I/O
    cin.tie(nullptr);            // Untie cin from cout for faster input

    int t;
    cin >> t;

    while (t--) {
        int n, m, L;
        cin >> n >> m >> L;

        vector<pair<int, int>> I(n); // To store "l r" pairs
        for (auto& [l, r] : I) cin >> l >> r;

        vector<pair<int, int>> P(m); // To store "x p" pairs
        for (auto& [x, p] : P) cin >> x >> p;

        sort(P.begin(), P.end()); // Sort P based on x to process in order

        priority_queue<int> heap; // Max heap (default is max in C++)
        int j = 1, c = 0;         // Initial jumping power and consumed powerups
        size_t idx = 0;           // Index to iterate over P

        for (auto [l, r] : I) {
            // Add powerups at positions before l to the heap
            while (idx < P.size() && P[idx].first < l) {
                heap.push(P[idx].second);
                idx++;
            }

            // Increase the jumping power until we can reach r
            while (j < r - l + 2) {
                if (heap.empty()) {
                    j = -1;
                    break; // Abort
                }
                j += heap.top();
                heap.pop();
                c++;
            }

            if (j == -1) break; // Abort if jumping power is insufficient
        }

        cout << (j >= 0 ? c : -1) << '\n';
    }

    return 0;
}

File: r988/D.py
import heapq

for _ in range(int(input())):
    n, m, L = map(int, input().split())

    # Read n "l r" integer pairs into a list
    I = [list(map(int, input().split())) for _ in range(n)]

    # Read m "x p" integer pairs into a list
    P = [list(map(int, input().split())) for _ in range(m)]

    # Heap of powerups, initial jumping power, and consumed powerups
    p, j, c, idx = [], 1, 0, 0

    for (l, r) in I:
        # Add powerups at positions before l to the heap
        while idx < len(P) and P[idx][0] < l:
            heapq.heappush(p, -P[idx][1])
            idx += 1
        
        # Increase the jumping power until we can reach r
        while j < r - l + 2:
            if not p:
                j = -1
                break # abort
            j -= heapq.heappop(p)
            c += 1
        
        if j == -1: break # abort

    print(c if j >= 0 else -1)